maro/simulator/scenarios/finance/business_engine.py

```diff
@@ -0,0 +1,653 @@
+
+    def take_action(self, action: Action, tick: int) -> TradeResult:
+        # 1. can trade -> bool
+        # 2. return (stock, sell/busy, stock_price, number, tax)
+        # 3. update stock.account_hold_num
+        rets = []
+        if action.state == ActionState.pending:
+            # not canceled
+            stock = self._stocks[action.item]
+            is_trigger = action.is_trigger(stock.opening_price, stock.trade_volume)
+            if is_trigger:
+                print(f"  <<>>  executing {action}")
+                if not self._allow_split:
+                    rets = self.trade(action)
+                else:
+                    # TODO: Split trade should be implemented
+                    rets = self.split_trade(action)
+
+                if len(rets) > 0:
+                    action.state = ActionState.success
+                    action.finish_tick = tick
+                    for ret in rets:
+                        print(f"  <<>>  trading: {ret}")
+                        if action.direction == OrderDirection.buy:
+                            self._stocks[action.item].average_cost = (
+                                (self._stocks[action.item].account_hold_num * self._stocks[action.item].average_cost) +
+                                (ret.price_per_item * ret.trade_number)
+                            ) / (self._stocks[action.item].account_hold_num + ret.trade_number)
+                            # day trading is considered
+                            if self._conf["trade_constraint"]["allow_day_trade"]:
+                                self._stocks[action.item].account_hold_num += ret.trade_number
+                            else:
+                                if action not in self._executing_orders:
+                                    self._executing_orders.append(action)
+                        else:
+                            self._stocks[action.item].account_hold_num -= ret.trade_number
+                else:
+                    action.state = ActionState.failed
+                    action.finish_tick = tick
+            else:
+                if action.life_time > 1:
+                    action.life_time -= 1
+                    # move to _on_action_recieved
+                    # self._event_buffer.gen_atom_event(tick + 1, DecisionEvent, action)
+                    if tick + 1 not in self._pending_orders:
+                        self._pending_orders[tick + 1] = []
+                    self._pending_orders[tick + 1].append(action)
+                else:
+                    action.state = ActionState.expired
+                    action.finish_tick = tick
+            action.action_result = rets
+
+        else:
+            # order canceled
+            pass
+        return rets
+
+    def cancel_order(self, action: Action, tick: int):
+        action.action.state = ActionState.canceled
+        action.action.finish_tick = tick
+        action.action.life_time = 0
```

Comment:
why action.action.life_time = 0 here?

what's the action.life_time ?

Reply:
I finally get the meaning of the life_time, I think this naming is confusing, change it to sth like ``remaining_life_time``?

Reply:
ok